Initial commit

SimonBarendse opened the pull request.

SimonBarendse committed:
Initial commit

jpcoenen commented on webhook.go:
```diff
@@ -0,0 +1,178 @@
+		return false, nil
+	}
+
+	version, enabled := pod.Annotations["secrethub"]
```
Is this annotation specific/clear enough? I feel like we're leaving very little wiggle-room for making changes in the future.

For example, what I could imagine is something like:
```
secrethub/inject: true
secrethub/version: v0.1.2.3
```
Or
```
secrethub/inject/enabled: true
secrethub/inject/version: v0.1.2.3
```

This also allows leaving out the version, in which case the latest version is pulled.

florisvdg replied:
Or:
```
secrethub/inject/myapp: true
secrethub/inject/nginx: false
```

That would solve this https://github.com/secrethub/secrethub-kubernetes-mutating-webhook/pull/1/files#r435796044

Because we can then leave out this check:

https://github.com/secrethub/secrethub-kubernetes-mutating-webhook/blob/a89f26ee2f62e17be2017bbae85911580be4570f/webhook.go#L118

SimonBarendse replied:
Yes, allowing for more wiggle-room would make it easier/possible to keep backwards-compatibility in the future. :+1: I'll think about a way to achieve this.

> This also allows leaving out the version, in which case the latest version is pulled.

Is this a use case we want to support or encourage?

I think I'd prefer the idempotency over automatically updating to the latest version. Especially since automatically updating would mean we can never do a major release of the CLI that includes breaking changes, without breaking deployments. Even though we've kept the CLI backwards compatible for a long time now, I'd prefer to keep the option open to ever do a major release. Eventually, there might come a time when we'll need this.
Having the annotation with the version allows users to time the updates of the CLI.

SimonBarendse replied:
Given the scenario of using external factors, such as a `secrethub.env` file you've described below, we may want to enable the mutation on a per-container basis instead of per-pod. What do you think?

SimonBarendse replied:
I like your proposal @florisvdg :+1:

Only thing I'd change is `inject`, as this may be confused with the behavior of the `secrehtub inject` command. How about `secrethub/run/my-app: true`?

jpcoenen replied:
```
secrethub/inject/myapp: true
secrethub/inject/nginx: false
```
I don't know if varying annotation names are idiomatic for k8s. Do you have an example of this happening somewhere else?

---

We could also do something like:
```
secrethub/inject/include: myapp
```
And
```
secrethub/inject/exclude: nginx
```

Where the default is "run on all containers", unless one of these two is specified, then it's :"all but listed" or "only listed".

florisvdg replied:
> I don't know if varying annotation names are idiomatic for k8s

Good point.

> Only thing I'd change is inject, as this may be confused with the behavior of the `secrethub inject` command. How about `secrethub/run/my-app: true`?

True, alternative: 
```
secrethub/mutate/include: myapp1,myapp2
```

Or keep it short with just 1 verb:
```
secrethub/mutate: myapp1,myapp2
```

SimonBarendse replied:
Okay, I'll add:
```
secrethub/mutate: myapp1,myapp2
secrethub/version: 0.38.0
```

This means the same SecretHub CLI version is used for all containers in the same pod. I think that's acceptable and outweighs having to configure the SecretHub version for every container.

If we ever want to add more control in the future, we can consider accepting the following syntax as well:
```
secrethub/mutate: myapp1,myapp2
secrethub/version: myapp1:0.38.0,myapp2:0.39.0
```
For now, I'd skip this complexity. It can be added in a backwards compatible way.

jpcoenen commented on webhook.go:
```diff
@@ -0,0 +1,178 @@
+// command if it contains SecretHub references.
+func (m *SecretHubMutator) mutateContainer(_ context.Context, c *corev1.Container) (*corev1.Container, bool, error) {
+	// Ignore if there are no SecretHub references in the container.
+	if !m.hasSecretHubReferences(c.Env) {
+		return c, false, nil
+	}
```
What if a container has a `secrethub.env` file (maybe provided by a volume)?

SimonBarendse replied:
Interesting use case. We've considered the use of `secrethub.env` files in Kubernetes before, but figured that in that case you'd include the `secrethub.env` file in the image and therefore it would make sense to include the CLI in the image as well. I have not yet considered including the `secrethub.env` file via a volume. How do you imagine this would work? Where would the `secrethub.env` file live? What are the advantages of having the `secrethub.env` file over configuring the environment in the kube files?

jpcoenen replied:
Well. The main thing I am hinting at: do we want to do this automatic behaviour if someone has explicitly requested injection of the CLI through an annotation?

SimonBarendse replied:
I'm not sure.

Currently, you're annotating a pod to indicate that SecretHub should be attached to a container when secret references are used for that container. When a pod contains multiple containers, you may not want to include SecretHub for every container.

We could move the annotation to container level. This gives the user more control at the cost of having to annotate multiple times if multiple containers within the pod use SecretHub.

SimonBarendse replied:
Hmm, it appears you cannot annotate containers.

SimonBarendse replied:
I'm only now seeing @florisvdg suggestion here: https://github.com/secrethub/secrethub-kubernetes-mutating-webhook/pull/1#discussion_r435876446

That would solve annotating on container level.

jpcoenen commented on webhook.go:
```diff
@@ -0,0 +1,178 @@
+
+// Mutate implements MutateFunc and provides the top-level entrypoint for object
+// mutation.
+func (m *SecretHubMutator) Mutate(ctx context.Context, obj metav1.Object) (bool, error) {
```
Can we be sure that this is only ran once for every Pod?

SimonBarendse replied:
I believe that within a single request, `Mutate` is [called only  once](https://github.com/slok/kubewebhook/blob/v0.10.0/pkg/webhook/mutating/webhook.go#L128).

And because mutation starts from the pod specification that is sent with the request, it is idempotent between failed requests. Sending the same request twice doesn't lead to a different response.

Why do you ask?

jpcoenen replied:
I aked because I've also seen mutating webhooks that store in an annotation whether they have executed. So I wanted to make sure this thing is idempotent. But if you're sure that it is always called with the original spec as input, it sounds like it is OK.

SimonBarendse replied:
Ah. I've seen a similar [example](https://github.com/slok/kubewebhook/tree/v0.10.0/examples/pod-annotate). I believe this annotation is an example and a way to show the mutator has done something and is not part of any logic that provides idempotency.

jpcoenen commented on README.md:
```diff
@@ -1 +1,11 @@
+1. It mounts the volume to the target container.
+1. It prefixes the target containers command with `<path/to/volume>/secrethub run --`.
+
+This project is based on and heavily inspired by [Berglas' Kubernetes Mutating Webhook](https://github.com/GoogleCloudPlatform/berglas/tree/v0.5.1/examples/kubernetes).
```
```suggestion
This project is based on and heavily inspired by [Berglas's Kubernetes Mutating Webhook](https://github.com/GoogleCloudPlatform/berglas/tree/v0.5.1/examples/kubernetes).
```

Leaving out the `s` is a Dutch thingy :wink:

jpcoenen reviewed (commented):
Coolio!

SimonBarendse committed:
Fix typo in README

Co-authored-by: Joris Coenen <jpcoenen@users.noreply.github.com>

SimonBarendse committed:
Split annotation in enabling on specific containers and version

- This allows to leave out the version to default to latest.
- This allows to indicate which containers to run on and which
  not to run on.

SimonBarendse committed:
Also mutate containers if they don't have secret references

If the container is explicitely configured to be mutated, mutate
it also if there are no secret references in the environment of the
pod spec. This allows for other ways to use secrethub run. For
example, you could supply a secrethub.env file over a volume or
you could set secret reference environment variables in some other
way.

SimonBarendse committed:
Merge branch 'feature/initial-commit' of github.com:secrethub/secrethub-kubernetes-mutating-webhook into feature/initial-commit

florisvdg commented on webhook.go:
```diff
@@ -54,15 +54,31 @@ func (m *SecretHubMutator) Mutate(ctx context.Context, obj metav1.Object) (bool,
 
+	containers := map[string]struct{}{}
+
+	for _, container := range strings.Split(containersStr, ",") {
```
Let's also allow `true`.

jpcoenen replied:
Then what is the behaviour if you have a container named `true`? I think if we want the easy behaviour of "just mutate all containers", we'd have to stick to seperate annotations:

All containers:
```
secrethub/mutate: true
```

Only `myapp1`:
```
secrethub/mutate: true
secrethub/include: myapp1
```

All containers but `myapp1`:
```
secrethub/mutate: true
secrethub/exclude: myapp1
```

SimonBarendse replied:
Do we need to support naming your container `true` and wanting to enable mutation on just that container? That doesn't sound like a realistic scenario at all and I think we can justify having to change your containers name in that case.

That being said, what's the advantage of allowing `true` to enable on all containers? Although it's nice that there's no dependency on the name anymore, I think depending on the name is quite common in Kubernetes, so name changes will probably not taken up lightly anyways? When mutation is enabled for all containers, you might not think about this when adding a new container to the spec. Do you want mutation to be enabled by default in that case? Do note that all containers that are mutated require the command to be explicitly defined in the containers spec.

Furthermore, do note that we have removed the check for secret references, to make it possible to enable mutation for other use cases (e.g. `secrethub.env` file in a volume), so setting mutation to be enabled by default for any containers that are added to the spec also affects containers that do not use SecretHub.

We could introduce an option `secrethub/mutate: detect` to only enable it when secret references are used.

Given all of the above, I'm leaning towards the simple set up where you opt-in for each container explicitly.

jpcoenen replied:
> Do we need to support naming your container true and wanting to enable mutation on just that container? That doesn't sound like a realistic scenario at all and I think we can justify having to change your containers name in that case.

I do not agree. It shows that the annotation is being abused. Sure, there is probably no one naming their containers `true`, but even having to discuss this point shows that something is smelly.

The same can be said for `secrethub/mutate: detect`.

> depending on the name is quite common in Kubernetes, so name changes will probably not taken up lightly anyways?

Do you have any examples/references for this? If this is indeed the case, I am totally fine with the current situation.

SimonBarendse replied:
Yeah, some examples:
volume names => volumeMounts.name
secret names => env.valueFrom.SecretKeyRef
secret names => volumes.secret.secretName

SimonBarendse replied:
> even having to discuss this point shows that something is smelly

Yeah, I do agree with you there. It doesn't feel completely right.

jpcoenen replied:
From the API spec:

> Name of the container specified as a DNS_LABEL. Each container in a pod must have a unique name (DNS_LABEL). Cannot be updated.

Combined with you answers, sounds reasonable to use that as a selector.

jpcoenen reviewed (commented):
I have one open comment: https://github.com/secrethub/secrethub-kubernetes-mutating-webhook/pull/1#discussion_r436583663.
